### DISCUSSION TITLE
Any firmware disassembler

### DISCUSSION THREAD

**Discussion by palainp** [arendst/Tasmota#19766] [Q&A] [answered]
I have from times to times unwanted restarts. I cannot use the default firmware as I need a specific driver so I have to compile it myself with the latest repo code. Thanks to the helpful discord I can see that the reboot is caused to Exception:0, the various addresses after that are not the same from one reboot to another, but there's always a common walling path: ` "CallChain":[... , "40240c31","40100c00","40250358","40102019"]}`.
Probably those are from a mainloop on something like that, my device going mad itself and it's not a software bug, but to rule that out, is there a disassembler I can use to find what these addresses are related with?

**Comment by sfromis** [comment]
Better ask Google, something like *ESP8266 disassembler*

But first, it may be relevant to consider if it even plausibly is a software issue, as especially power issues are *common* causes of exceptions. And to use the available symbol-based decoding. Disassembling is not really recommended for debugging.

**Comment by lalo-uy** [comment]
More than a disassembler you may wato to look to linker map. There are the
addresses of all the functions.

**Reply by barbudor** [reply]
When you compile your own firmware, the binary is is Tasmota/build_output/firmware/ and the map file is in Tasmota/build_output/map/

**Comment by Staars** [comment]
Additionally you can uncomment this line to get a diassembly with every build:
https://github.com/arendst/Tasmota/blob/d05d76b4812dd391e3b366814bd8ac9c3c49bb51/platformio.ini#L80

**Comment by s-hadinger** [comment, marked as answer]
It's more straightforward to use the `.map` file as indicated above.

However if you really want to disassemble (which is very interesting per se):

`~/.platformio/packages/toolchain-xtensa-esp32/bin/xtensa-esp32-elf-objdump -S .pio/build/tasmota/firmware.elf  -S .pio/build/tasmota/firmware.elf > disassemble.txt`

You need to adapt to the appropriate toolchain whether ESP8266 or ESP32

**Comment by palainp** [comment]
Thanks all for your suggestions (it's the esp8266 version)! It's indeed probably an hardware issue, but I wanted to rule out something related to anything I could have done inside the code.

So far I tracked down the last two reboots inside:
```C
void SleepDelay(uint32_t mseconds) {
  if (!TasmotaGlobal.backlog_nodelay && mseconds) {
    uint32_t wait = millis() + mseconds;
    while (!TimeReached(wait) && !Serial.available()) {  // We need to service serial buffer ASAP as otherwise we get uart buffer overrun
      XdrvXsnsCall(FUNC_SLEEP_LOOP);  // Main purpose is reacting ASAP on serial data availability or interrupt handling (ADE7880)
      delay(1);
    }
  } else {
    delay(0);
  }
}
```
One time on the `XdrvXsnsCall(.)` call, the other time on the `delay(1)` call.
I'm not sure why is called this function, and I'll continue the hunt :)

**Reply by sfromis** [reply]
To me that looks like where Tasmota spends most of it's time, meaning that hardware-related issues are quite likely to hit there.

**Comment by palainp** [comment]
As pointed by @sfromis , I didn't get any further on that issue,  concluded to something hardware related and changed the shutter.
Anyway the path was interresting, thanks everyone for your indications !
